src/charm.py

```diff
@@ -0,0 +1,37 @@
+#!/usr/bin/env python3
+# Copyright 2024 dima.tisnek@canonical.com
+# See LICENSE file for licensing details.
+"""Charmed Gubernator."""
+
+import logging
+
+import ops
+
+from charms.traefik_k8s.v2.ingress import IngressPerAppRequirer
+
+# Log messages can be retrieved using juju debug-log
+logger = logging.getLogger(__name__)
+
+VALID_LOG_LEVELS = ["info", "debug", "warning", "error", "critical"]
+
+
+class HexanatorCharm(ops.CharmBase):
+    """Charm the service."""
+
+    def __init__(self, *args):
+        super().__init__(*args)
```

Comment:
For what it's worth, we're encouraging people to have this more explicit, like so:

```suggestion
    def __init__(self, framework: ops.Framework):
        super().__init__(framework)
```

And then in the observe calls to just say `framework.observe` rather than `self.framework.observe`.

Obviously, it'll work either way - Ben and I felt that the variable `args` wasn't really gaining anything and that the explicit version was clearer to understand (especially since you no longer need to know that the super call will set `self.framework`).

Reply:
Done